File: frontend/src/lib/api/bandwidth.ts
import api from './client';

export interface BandwidthReport {
    room_id: string;
    session_id: string;
    bytes_sent: number;
    bytes_received: number;
    packets_sent: number;
    packets_lost: number;
    duration: number;
}

export interface UserBandwidthStats {
    user_id: string;
    total_bytes_sent: number;
    total_bytes_received: number;
    total_duration: number;
    meeting_count: number;
}

export interface BandwidthRecord {
    id: string;
    user_id: string;
    room_id: string;
    session_id: string;
    bytes_sent: number;
    bytes_received: number;
    packets_sent: number;
    packets_lost: number;
    duration: number;
    created_at: string;
    updated_at: string;
}

export const bandwidthApi = {
    reportBandwidth: async (data: BandwidthReport): Promise<void> => {
        await api.post('/bandwidth/report', data);
    },

    getStats: async (): Promise<UserBandwidthStats> => {
        const response = await api.get<UserBandwidthStats>('/bandwidth/stats');
        return response.data;
    },

    getHistory: async (limit: number = 10): Promise<BandwidthRecord[]> => {
        const response = await api.get<BandwidthRecord[]>(`/bandwidth/history?limit=${limit}`);
        return response.data;
    },
};

File: backend/internal/core/domain/bandwidth/bandwidth.go
package bandwidth

import (
	"context"
	"time"
)

// BandwidthRecord represents a bandwidth usage report for a user in a room
type BandwidthRecord struct {
	ID            string    `json:"id" bson:"_id"`
	UserID        string    `json:"user_id" bson:"user_id"`
	RoomID        string    `json:"room_id" bson:"room_id"`
	SessionID     string    `json:"session_id" bson:"session_id"`
	BytesSent     int64     `json:"bytes_sent" bson:"bytes_sent"`
	BytesReceived int64     `json:"bytes_received" bson:"bytes_received"`
	PacketsSent   int64     `json:"packets_sent" bson:"packets_sent"`
	PacketsLost   int64     `json:"packets_lost" bson:"packets_lost"`
	Duration      int64     `json:"duration" bson:"duration"` // duration in seconds
	CreatedAt     time.Time `json:"created_at" bson:"created_at"`
	UpdatedAt     time.Time `json:"updated_at" bson:"updated_at"`
}

// UserBandwidthStats represents aggregated bandwidth statistics for a user
type UserBandwidthStats struct {
	UserID             string `json:"user_id" bson:"_id"`
	TotalBytesSent     int64  `json:"total_bytes_sent" bson:"total_bytes_sent"`
	TotalBytesReceived int64  `json:"total_bytes_received" bson:"total_bytes_received"`
	TotalDuration      int64  `json:"total_duration" bson:"total_duration"`
	MeetingCount       int    `json:"meeting_count" bson:"meeting_count"`
}

// Repository interface defines methods for bandwidth data access
type Repository interface {
	// Create inserts a new bandwidth record
	Create(ctx context.Context, record *BandwidthRecord) error

	// Update updates an existing bandwidth record
	Update(ctx context.Context, record *BandwidthRecord) error

	// FindBySession finds a bandwidth record for a specific user, room and session
	FindBySession(ctx context.Context, userID, roomID, sessionID string) (*BandwidthRecord, error)

	// GetUserStats calculates aggregated statistics for a user
	GetUserStats(ctx context.Context, userID string) (*UserBandwidthStats, error)

	// GetUserHistory retrieves bandwidth usage history for a user
	GetUserHistory(ctx context.Context, userID string, limit int) ([]*BandwidthRecord, error)
}
